src/PureScript/CST/ModuleGraph.purs

```diff
@@ -0,0 +1,117 @@
+    Tuple name (Set.fromFoldable (map getImportName imports))
+
+  getImportName (ImportDecl { "module": Name { name } }) = name
+
+data ModuleSort e
+  = Sorted (Array (ModuleHeader e))
+  | CycleDetected (Array (ModuleHeader e))
+
+sortModules :: forall e. Array (ModuleHeader e) -> ModuleSort e
+sortModules moduleHeaders = do
+  let
+    knownModuleHeaders :: Map ModuleName (ModuleHeader e)
+    knownModuleHeaders =
+      moduleHeaders
+        # map (\header@(ModuleHeader { name: Name { name } }) -> Tuple name header)
+        # Map.fromFoldable
+
+    graph = moduleGraph moduleHeaders
+    lookupModuleHeaders = Array.mapMaybe (flip Map.lookup knownModuleHeaders) <<< List.toUnfoldable
+
+  case bimap lookupModuleHeaders lookupModuleHeaders (topoSort graph) of
+    Left cycle -> CycleDetected cycle
+    Right sorted -> Sorted sorted
+
+type TopoSortArgs a =
+  { roots :: Set a
+  , sorted :: List a
+  , usages :: Map a Int
+  }
+
+topoSort :: forall a. Ord a => Graph a -> Either (List a) (List a)
+topoSort graph = do
+  _.sorted <$> go { roots: startingModules, sorted: Nil, usages: importCounts }
+  where
+  go :: TopoSortArgs a -> Either (List a) (TopoSortArgs a)
+  go { roots, sorted, usages } = case Set.findMin roots of
+    Nothing ->
+      if all (eq 0) usages then
+        Right { roots, sorted, usages }
+      else do
+        let
+          nonLeaf = Set.fromFoldable $ Map.keys $ Map.filterWithKey (\a count -> count /= 0 && Map.lookup a graph /= Nothing && Map.lookup a graph /= Just Set.empty) usages
+
+        case foldl (\b a -> if isJust b then b else runFree (un Identity) (un Compose (depthFirst { path: Nil, visited: Set.empty, curr: a }))) Nothing nonLeaf of
+          Just cycle -> Left cycle
+          Nothing -> Left Nil
+
+    Just curr -> do
+      let
+        reachable = fromMaybe Set.empty (Map.lookup curr graph)
+        usages' = foldl decrementImport usages reachable
+      go
+        { roots: foldl (appendRoots usages') (Set.delete curr roots) reachable
+        , sorted: Cons curr sorted
+        , usages: usages'
+        }
+
+  appendRoots :: Map a Int -> Set a -> a -> Set a
+  appendRoots usages roots curr = maybe roots (flip Set.insert roots) do
+    count <- Map.lookup curr usages
+    isRoot (Tuple curr count)
+
+  decrementImport :: Map a Int -> a -> Map a Int
+  decrementImport usages k = Map.insertWith add k (-1) usages
+
+  startingModules :: Set a
+  startingModules = Set.fromFoldable $ Map.keys $ Map.filterWithKey (\k v -> isJust (isRoot (Tuple k v))) importCounts
+
+  importCounts :: Map a Int
+  importCounts = Map.fromFoldableWith add do
+    Tuple a bs <- Map.toUnfoldable graph
+    [ Tuple a 0 ] <> map (flip Tuple 1) (Set.toUnfoldable bs)
+
+  isRoot :: Tuple a Int -> Maybe a
+  isRoot (Tuple a count) = if count == 0 then Just a else Nothing
+
+  depthFirst :: { path :: List a, visited :: Set a, curr :: a } -> Compose (Free Identity) Maybe (List a)
```

Comment:
This Free isn't doing anything to mitigate stack. Using Free to mitigate stack only works when recursive calls are binds, but you aren't doing that here. I would be very surprised if module dependency graphs were deep enough to blow the stack, so I'm inclined to just not bother until it becomes a problem.

Reply:
Yeah, not sure what I was thinking with all that. It'd have to be a pretty crazy cycle to blow the stack so maybe we can punt on it for now.